### PR TITLE
Fix issue where service is not shutdown when check is cancelled. Closes #1250

### LINKED ISSUE #1250: service is not shutdown when check is cancelled

### PR DESCRIPTION
**Pull request by binaek**

### CODE REVIEW

File: cmd/check.go [hunk 1]
```diff
@@ -138,7 +140,6 @@ func runCheckCmd(cmd *cobra.Command, args []string) {
 	}
 
 	// pull out useful properties
-	ctx := initData.ctx
 	workspace := initData.workspace
```
Comment (kaidaguerre) [inline cmd/check.go:143]: an reason not to use the ctx var to tidy things up?

File: control/controlexecute/control_run.go [hunk 2]
```diff
@@ -178,7 +178,10 @@ func (r *ControlRun) Execute(ctx context.Context, client db_common.Client) {
 	}
 	r.Lifecycle.Add("got_session")
 	dbSession := sessionResult.Session
-	defer dbSession.Close()
+	defer func() {
+		// do this in a closure, otherwise the argument will not get evaluated in calltime
```
Comment (kaidaguerre) [inline control/controlexecute/control_run.go:182]: interesting! I always forget this

File: db/db_common/db_session.go [hunk 3]
```diff
@@ -33,9 +35,22 @@ func (s *DatabaseSession) UpdateUsage() {
...
-func (s *DatabaseSession) Close() error {
+func (s *DatabaseSession) Close(waitForCleanup bool) error {
+	var err error
 	if s.Connection != nil {
-		err := s.Connection.Close()
+		if waitForCleanup {
+			s.Connection.Raw(func(driverConn interface{}) error {
+				conn := driverConn.(*stdlib.Conn)
+				select {
+				case <-time.After(5 * time.Second):
```
Comment (kaidaguerre) [inline db/db_common/db_session.go:45]: is 5 secs long enough?
what happens if we timeout? nothing? i.e. we just continue shutting down? That seems desirable

File: db/db_local/stop_services.go [hunk 4]
```diff
@@ -68,8 +69,8 @@ func ShutdownService(invoker constants.Invoker) {
 
 }
 
-// GetCountOfConnectedClients returns the number of clients currently connected to the service
-func GetCountOfConnectedClients(ctx context.Context) (i int, e error) {
+// GetCountOfThirdPartyClients returns the number of clients currently connected to the service
```
Comment (kaidaguerre) [inline db/db_local/stop_services.go:72]: update comment

File: db/db_client/db_client.go [hunk 5]
```diff
@@ -74,6 +78,15 @@ func establishConnection(ctx context.Context, connStr string) (*sql.DB, error) {
 	utils.LogTime("db_client.establishConnection start")
 	defer utils.LogTime("db_client.establishConnection end")
 
+	connConfig, _ := pgx.ParseConfig(connStr)
+	connConfig.PreferSimpleProtocol = true
```
Comment (kaidaguerre) [inline db/db_client/db_client.go:82]: what is this?

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

File: db/db_common/db_session.go (hunk 3 above)
Comment (binaek) [reply db/db_common/db_session.go:45]: for now, we just continue.

File: db/db_client/db_client.go (hunk 5 above)
Comment (binaek) [reply db/db_client/db_client.go]: experiment - fixed

File: db/db_local/stop_services.go (hunk 4 above)
Comment (binaek) [reply db/db_local/stop_services.go]: updated

File: control/controlexecute/control_run.go (hunk 2 above)
Comment (binaek) [reply control/controlexecute/control_run.go:182]: as do I 🥇

File: cmd/check.go (hunk 1 above)
Comment (binaek) [reply cmd/check.go]: tidied up

File: db/db_client/db_client.go [hunk 6]
```diff
@@ -74,6 +78,14 @@ func establishConnection(ctx context.Context, connStr string) (*sql.DB, error) {
 	utils.LogTime("db_client.establishConnection start")
 	defer utils.LogTime("db_client.establishConnection end")
 
+	connConfig, _ := pgx.ParseConfig(connStr)
+
```
Comment (kaidaguerre) [inline db/db_client/db_client.go:82]: remove lines
```
        connConfig, _ := pgx.ParseConfig(connStr)
	connConfig.RuntimeParams = map[string]string{
		"application_name": runtime_constants.PgClientAppName,
	}
	connStr = stdlib.RegisterConnConfig(connConfig)
```

Comment (kaidaguerre) [reply db/db_client/db_client.go:82]: also add comment - `set appName so we can track connections` or something

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

**Pull request merged**